### PR TITLE
Gcloud functions gen2 tests

### PR DESCRIPTION
**Pull request by penx**
Add unit tests to [gcloud-functions-gen2](https://github.com/eastlondoner/remix/tree/gcloud-functions-gen2) branch.

These are the Express unit tests, updated to use `getTestServer` and `http` from `@google-cloud/functions-framework` so that they behave as per google cloud functions (e.g. rawBody is present on request).

These require Jest@28 which is in PR at https://github.com/remix-run/remix/pull/3400, and can be manually updated to see the tests pass.

### CODE REVIEW

File: packages/remix-gcloud-functions-gen2/__tests__/server-test.ts [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+    createRequestHandler({
+      // We don't have a real app to test, but it doesn't matter. We
+      // won't ever call through to the real createRequestHandler
+      build: undefined,
+    })
+  );
+  return getTestServer("remixServer");
+}
+
+describe("express createRequestHandler", () => {
```
Comment (MichaelDeBoey) [inline packages/remix-gcloud-functions-gen2/__tests__/server-test.ts]:
```suggestion
describe("google-cloud-functions createRequestHandler", () => {
```

File: packages/remix-gcloud-functions-gen2/__tests__/server-test.ts [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+      expect(res.headers["set-cookie"]).toEqual([
+        "first=one; Expires=0; Path=/; HttpOnly; Secure; SameSite=Lax",
+        "second=two; MaxAge=1209600; Path=/; HttpOnly; Secure; SameSite=Lax",
+        "third=three; Expires=Wed, 21 Oct 2015 07:28:00 GMT; Path=/; HttpOnly; Secure; SameSite=Lax",
+      ]);
+    });
+  });
+});
+
+describe("express createRemixHeaders", () => {
```
Comment (MichaelDeBoey) [inline packages/remix-gcloud-functions-gen2/__tests__/server-test.ts]:
```suggestion
describe("google-cloud-functions createRemixHeaders", () => {
```

File: packages/remix-gcloud-functions-gen2/__tests__/server-test.ts [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+            "__other=some_other_value; Path=/; Secure; HttpOnly; MaxAge=3600; SameSite=Lax",
+          ],
+          Symbol(context): null,
+        }
+      `);
+    });
+  });
+});
+
+describe("express createRemixRequest", () => {
```
Comment (MichaelDeBoey) [inline packages/remix-gcloud-functions-gen2/__tests__/server-test.ts]:
```suggestion
describe("google-cloud-functions createRemixRequest", () => {
```

File: packages/remix-gcloud-functions-gen2/jest.config.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+/** @type {import('@jest/types').Config.InitialOptions} */
+module.exports = {
+  ...require("../../jest/jest.config.shared"),
+  displayName: "gcloud-functions-gen2",
```
Comment (MichaelDeBoey) [inline packages/remix-gcloud-functions-gen2/jest.config.js]:
```suggestion
  displayName: "google-cloud-functions",
```

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]

**Comment by eastlondoner** [comment]
@penx I think we need to add a devDependency on `@google-cloud/functions-framework` so the tests can run - CI fails with 

Is there anything these tests are doing that specifically requires `Jest@28` ?

**Comment by penx** [comment]
> Is there anything these tests are doing that specifically requires Jest@28 ?

```ts
import { getTestServer } from "@google-cloud/functions-framework/testing";
```

This import makes use of [package exports](https://nodejs.org/api/packages.html#exports):

https://github.com/GoogleCloudPlatform/functions-framework-nodejs/blob/759aac18ad297363c4651efcf19ab2bf8fa9625a/package.json#L13

Full support for this [was added in Jest 28](https://jestjs.io/blog/2022/04/25/jest-28#packagejson-exports).

In Jest 27,  server-test.ts will fail saying it cannot find `@google-cloud/functions-framework/testing`.

**Comment by penx** [comment]
> I think we need to add a devDependency on @google-cloud/functions-framework so the tests can run - CI fails with

It's in the devDependencies for the package already

https://github.com/penx/remix/blob/689da9c07f674be05d3b802b83afaa8aa3bd34bb/packages/remix-gcloud-functions-gen2/package.json#L20-L22

And as per the comment at https://github.com/remix-run/remix/pull/2266#discussion_r889528068, this should be moved to dependencies.

This should be sufficient - the two remaining issues are that I didn't update the lock file in this PR and that Jest 27 will complain about a missing export as per my other comment.

**Comment by MichaelDeBoey** [comment]
@penx Now that https://github.com/GoogleCloudPlatform/functions-framework-nodejs/pull/461 is merged, do we still need `jest` v28?

**Comment by penx** [comment]
@MichaelDeBoey as discussed on discord, yes, so that the `/testing` import is supported

**Comment by eastlondoner** [comment]
> > I think we need to add a devDependency on @google-cloud/functions-framework so the tests can run - CI fails with
> 
> It's in the devDependencies for the package already
> 
> https://github.com/penx/remix/blob/689da9c07f674be05d3b802b83afaa8aa3bd34bb/packages/remix-gcloud-functions-gen2/package.json#L20-L22
> 
> And as per the comment at [remix-run#2266 (comment)](https://github.com/remix-run/remix/pull/2266#discussion_r889528068), this should be moved to dependencies.
> 
> This should be sufficient - the two remaining issues are that I didn't update the lock file in this PR and that Jest 27 will complain about a missing export as per my other comment.

OK, currently this is failing with being unable to import though...
```
Summary of all failing tests
[83](https://github.com/eastlondoner/remix/runs/6762836078?check_suite_focus=true#step:6:84)
FAIL packages/remix-gcloud-functions-gen2/__tests__/server-test.ts
[84](https://github.com/eastlondoner/remix/runs/6762836078?check_suite_focus=true#step:6:85)
  ● Test suite failed to run
[85](https://github.com/eastlondoner/remix/runs/6762836078?check_suite_focus=true#step:6:86)
[86](https://github.com/eastlondoner/remix/runs/6762836078?check_suite_focus=true#step:6:87)
    Cannot find module '@google-cloud/functions-framework/testing' from '__tests__/server-test.ts'
[87](https://github.com/eastlondoner/remix/runs/6762836078?check_suite_focus=true#step:6:88)
[88](https://github.com/eastlondoner/remix/runs/6762836078?check_suite_focus=true#step:6:89)
       8 | import { http } from "@google-cloud/functions-framework";
[89](https://github.com/eastlondoner/remix/runs/6762836078?check_suite_focus=true#step:6:90)
       9 | // @ts-ignore
[90](https://github.com/eastlondoner/remix/runs/6762836078?check_suite_focus=true#step:6:91)
    > 10 | import { getTestServer } from "@google-cloud/functions-framework/testing";
[91](https://github.com/eastlondoner/remix/runs/6762836078?check_suite_focus=true#step:6:92)
         | ^
[92](https://github.com/eastlondoner/remix/runs/6762836078?check_suite_focus=true#step:6:93)
      11 |
[93](https://github.com/eastlondoner/remix/runs/6762836078?check_suite_focus=true#step:6:94)
      12 | import {
[94](https://github.com/eastlondoner/remix/runs/6762836078?check_suite_focus=true#step:6:95)
      13 |   createRemixHeaders,
[95](https://github.com/eastlondoner/remix/runs/6762836078?check_suite_focus=true#step:6:96)
[96](https://github.com/eastlondoner/remix/runs/6762836078?check_suite_focus=true#step:6:97)
      at Resolver.resolveModule (../../node_modules/jest-resolve/build/resolver.js:324:11)
[97](https://github.com/eastlondoner/remix/runs/6762836078?check_suite_focus=true#step:6:98)
      at Object.<anonymous> (__tests__/server-test.ts:10:1)
```

**Comment by penx** [comment]
Yes that is the jest issue that requires updating to v28 to support package.json exports

**Comment by eastlondoner** [comment]
Thanks @penx ! I cherry-picked the relevant commits from this PR into https://github.com/eastlondoner/remix/tree/gcloud-functions-gen2 after I rebased on the `dev` branch (to save you needing to rebase this).

Closing this PR because it's no longer needed after that cherry-pick.

**Pull request closed without merging**